common-rest/src/main/java/org/uniprot/api/common/repository/stream/store/uniprotkb/TaxonomyLineageServiceImpl.java

```diff
@@ -0,0 +1,162 @@
+import org.uniprot.store.config.searchfield.common.SearchFieldConfig;
+import org.uniprot.store.config.searchfield.factory.SearchFieldConfigFactory;
+import org.uniprot.store.config.searchfield.model.SearchFieldItem;
+import org.uniprot.store.search.document.taxonomy.TaxonomyDocument;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * @author jluo
+ * @date: 16 Oct 2019
+ */
+@Slf4j
+@Service
+public class TaxonomyLineageServiceImpl extends BasicSearchService<TaxonomyDocument, TaxonomyEntry>
+        implements TaxonomyLineageService {
+    private final SearchFieldConfig searchFieldConfig;
+
+    @Autowired
+    public TaxonomyLineageServiceImpl(TaxonomyLineageRepository taxRepo) {
+        super(
+                taxRepo,
+                new TaxonomyEntryConverter(),
+                new TaxonomySortClause(),
+                getQueryBoosts(),
+                null);
+        this.searchFieldConfig =
+                SearchFieldConfigFactory.getSearchFieldConfig(UniProtDataType.TAXONOMY);
+    }
+
+    @Cacheable("taxonomyCache")
+    @Override
+    public TaxonomyEntry findByUniqueId(String uniqueId) {
+        return super.findByUniqueId(uniqueId);
+    }
+
+    @Override
+    protected SearchFieldItem getIdField() {
+        return searchFieldConfig.getSearchFieldItemByName("id");
+    }
+
+    @Override
+    protected UniProtQueryProcessorConfig getQueryProcessorConfig() {
+        return UniProtQueryProcessorConfig.builder().build();
+    }
+
+    @Override
+    public TaxonomyEntry findById(long taxId) {
+        return findByUniqueId(String.valueOf(taxId));
+    }
+
+    @Override
+    public Map<Long, List<TaxonomyLineage>> findByIds(Set<Long> taxIds) {
+        String idFieldName = getIdField().getFieldName() + ":";
+        String query =
+                idFieldName
+                        + taxIds.stream()
+                                .map(String::valueOf)
+                                .collect(Collectors.joining(" OR " + idFieldName));
+        StreamRequest request = new TaxonomyLineageStreamRequest(query);
+        return this.stream(request)
+                .collect(Collectors.toMap(TaxonomyEntry::getTaxonId, TaxonomyEntry::getLineages));
+    }
+
+    private static SolrQueryConfig getQueryBoosts() {
+        return SolrQueryConfig.builder().queryFields("id").build();
+    }
+
+    static class TaxonomyEntryConverter implements Function<TaxonomyDocument, TaxonomyEntry> {
+
+        private final ObjectMapper objectMapper;
+
+        public TaxonomyEntryConverter() {
+            objectMapper = TaxonomyJsonConfig.getInstance().getFullObjectMapper();
+        }
+
+        @Override
+        public TaxonomyEntry apply(TaxonomyDocument taxonomyDocument) {
+            log.info("fetch taxonomy=" + taxonomyDocument.getTaxId());
+            try {
+                return objectMapper.readValue(
+                        taxonomyDocument.getTaxonomyObj(), TaxonomyEntry.class);
```

Comment:
is this fast?

Reply:
We use it to deserialise binary data from Solr.
This is a good point, we did not test other options for this. We just used the same serialisation method that we use for Voldemort. Do you think we should explorer other options? any suggestion? Would you mind open a Jira for it?